### PR TITLE
feat(sink): rewind log reader to isolate sink err

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As titled.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/common/log_store_impl/kv_log_store/reader.rs [hunk 1]
```diff
@@ -43,6 +46,52 @@ use crate::common::log_store_impl::kv_log_store::serde::{
...
+    }
+
+    async fn rewind_delay(&mut self, truncate_offset: Option<TruncateOffset>) {
+        match (&self.last_rewind_truncate_offset, &truncate_offset) {
+            (Some(prev_rewind_truncate_offset), Some(truncate_offset)) => {
+                if truncate_offset > prev_rewind_truncate_offset {
+                    self.last_rewind_truncate_offset = Some(*truncate_offset);
+                    // Have new truncate progress before this round of rewind.
+                    // Reset rewind backoff
+                    self.backoff_policy = initial_rewind_backoff_policy();
```
Comment (tabVersion) [inline src/stream/src/common/log_store_impl/kv_log_store/reader.rs:81]: is it for another error that occurs when rewinding the prev one?

File: src/stream/src/common/log_store_impl/kv_log_store/reader.rs [hunk 2]
```diff
@@ -404,7 +453,8 @@ impl<S: StateStore> LogReader for KvLogStoreReader<S> {
         Ok(())
     }
 
-    async fn rewind(&mut self) -> LogStoreResult<bool> {
+    async fn rewind(&mut self) -> LogStoreResult<(bool, Option<Bitmap>)> {
```
Comment (tabVersion) [inline src/stream/src/common/log_store_impl/kv_log_store/reader.rs:456]: whats the bitmap for?

File: src/stream/src/common/log_store_impl/kv_log_store/reader.rs (hunk 1 above)
Comment (wenym1) [reply src/stream/src/common/log_store_impl/kv_log_store/reader.rs:81]: No. The current logic is that, if the rewind is caused by the same error in the external sink, we will keep rewinding with exponential backoff delay. If the rewind is caused by a new error, we will reset to the initial delay.

File: src/stream/src/common/log_store_impl/kv_log_store/reader.rs (hunk 2 above)
Comment (wenym1) [reply src/stream/src/common/log_store_impl/kv_log_store/reader.rs:456]: Currently the initial vnode bitmap of sink writer is passed by sink writer param during `new`. Later the sink may update to new vnode bitmap, in which case the latest vnode bitmap will be different to the one in the initial sink writer param. So in rewind we will return the latest vnode bitmap from the log reader, and later when we create the sink writer, we pass the latest vnode bitmap.

File: src/connector/src/sink/mod.rs [hunk 3]
```diff
@@ -303,9 +304,40 @@ pub trait Sink: TryFrom<SinkParam, Error = SinkError> {
     }
 }
 
+pub trait SinkLogReader: Send + Sized + 'static {
```
[the remark this answers is not in the document]
Comment (wenym1) [reply src/connector/src/sink/mod.rs:307]: Yes, it's a subset. 

It is introduced to limit the methods called by the log sinker. Log sinkers are allowed to call only the `next_item` and `truncate`, and in this way, we can prohibit the illegal call on `init` and `rewind`.

File: src/stream/src/executor/sink.rs [hunk 4]
```diff
@@ -341,7 +340,13 @@ impl<F: LogStoreFactory> SinkExecutor<F> {
             })
             .monitored(metrics);
 
-        if let Err(e) = log_sinker.consume_log_and_sink(log_reader).await {
+        log_reader.init().await?;
+
+        while let Err(e) = sink
+            .new_log_sinker(sink_writer_param.clone())
```
[the remark this answers is not in the document]
Comment (wenym1) [reply src/stream/src/executor/sink.rs:346]: Yes. The `consume_log_and_sink` method of log sinker will take its ownership, so we need to create a new one. In practice, some log sinker will have some initial state in `new`, so we should rebuild a new one.

File: src/stream/src/common/log_store_impl/kv_log_store/reader.rs [hunk 5]
```diff
@@ -404,7 +453,8 @@ impl<S: StateStore> LogReader for KvLogStoreReader<S> {
...
+        self.rewind_delay.rewind_delay(self.truncate_offset).await;
```
[the remark this answers is not in the document]
Comment (wenym1) [reply src/stream/src/common/log_store_impl/kv_log_store/reader.rs:457]: It's a general practice for retrying after error. If the error is caused by crash of external sink and we don't have any backoff delay, we will keep repeatedly rewinding due to the same error in a crash loop, which may consume many CPU.

File: src/stream/src/executor/sink.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/stream/src/executor/sink.rs:346]: +1 to rebuild a new one with clean memory state.

File: src/stream/src/common/log_store_impl/kv_log_store/reader.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (StrikeW) [reply src/stream/src/common/log_store_impl/kv_log_store/reader.rs:457]: Got it, thanks.

**Review by StrikeW — APPROVED** [review]
LGTM

**Review by tabVersion — APPROVED** [review]
thanks, LGTM

File: src/stream/src/common/log_store_impl/kv_log_store/reader.rs [hunk 6]
```diff
@@ -43,6 +46,52 @@ use crate::common::log_store_impl::kv_log_store::serde::{
...
+
+impl RewindDelay {
+    fn new() -> Self {
+        Self {
+            last_rewind_truncate_offset: None,
+            backoff_policy: initial_rewind_backoff_policy(),
+        }
+    }
+
+    async fn rewind_delay(&mut self, truncate_offset: Option<TruncateOffset>) {
```
Comment (hzxa21) [inline src/stream/src/common/log_store_impl/kv_log_store/reader.rs:74]: It is better to add a metric to monitor retry delay and retry count or rate/

**Review by hzxa21 — APPROVED** [review]
LGTM. How do we test this feature? IMO, we can use madsim to inject failure or write an integration test for it.

**Comment by wenym1** [comment]
> LGTM. How do we test this feature? IMO, we can use madsim to inject failure or write an integration test for it.

Test added in https://github.com/risingwavelabs/risingwave/pull/14150.

**Pull request merged**